batch/batch/worker/worker.py

```diff
@@ -1789,6 +1796,323 @@ def __str__(self):
+    b = await read(reader, 4)
+    return struct.unpack('>i', b)[0]
+
+
+async def read_long(reader: asyncio.StreamReader) -> int:
+    b = await read(reader, 8)
+    return struct.unpack('>q', b)[0]
+
+
+async def read_bytes(reader: asyncio.StreamReader) -> bytes:
+    n = await read_int(reader)
+    return await read(reader, n)
+
+
+async def read_str(reader: asyncio.StreamReader) -> str:
+    b = await read_bytes(reader)
+    return b.decode('utf-8')
+
+
+@contextmanager
+def scoped_ensure_future(coro_or_future, *, loop=None) -> Iterator[asyncio.Future]:
+    fut = asyncio.ensure_future(coro_or_future, loop=loop)
+    try:
+        yield fut
+    finally:
+        fut.cancel()
+
+
+class BufferedOutputProcess:
+    @classmethod
+    async def create(cls, *args, **kwargs):
+        assert 'stdout' not in kwargs
+        assert 'stderr' not in kwargs
+
+        process = await asyncio.create_subprocess_exec(
+            *args, **kwargs, stdout=asyncio.subprocess.PIPE, stderr=asyncio.subprocess.PIPE
+        )
+        stop_event = asyncio.Event()  # FIXME: this seems unused
+        return cls(process, stop_event)
+
+    def __init__(self, process, stop_event: asyncio.Event):
+        self.process = process
+        self.stop_event = stop_event
+        self.buf = bytearray()
+        assert process.stdout is not None
+        self.stdout_pump = asyncio.ensure_future(self.pump_to_buffer(process.stdout))
+        assert process.stderr is not None
+        self.stderr_pump = asyncio.ensure_future(self.pump_to_buffer(process.stderr))
+
+    async def pump_to_buffer(self, strm: asyncio.StreamReader):
+        with scoped_ensure_future(self.stop_event.wait()) as stop_fut:
+            while not strm.at_eof() and not self.stop_event.is_set():
+                with scoped_ensure_future(strm.readline()) as read_fut:
+                    await asyncio.wait([read_fut, stop_fut], return_when=asyncio.FIRST_COMPLETED)
+                    if read_fut.done():
+                        result = read_fut.result()
+                        self.buf.extend(result)
+
+    def output(self) -> str:
+        return self.buf.decode()
+
+    def retrieve_and_clear_output(self) -> str:
+        buf = self.buf.decode()
+        self.buf = bytearray()
+        return buf
+
+    def kill(self):
+        return self.process.kill()
+
+    @property
+    def returncode(self) -> Optional[int]:
+        return self.process.returncode
+
+    def close(self):
+        self.kill()
+        self.stdout_pump.cancel()
+        self.stderr_pump.cancel()
+
+
+class JVM:
+    SPARK_HOME = find_spark_home()
+
+    FINISH_USER_EXCEPTION = 0
+    FINISH_ENTRYWAY_EXCEPTION = 1
+    FINISH_NORMAL = 2
+    FINISH_CANCELLED = 3
+    FINISH_JVM_EOS = 4
+
+    @classmethod
+    async def create_process(cls, socket_file: str) -> BufferedOutputProcess:
+        return await BufferedOutputProcess.create(
+            'java',
+            '-Xmx3500M',
```

Comment:
I'm a bit worried about the memory situation. How is memory shared amongst users? Should we require JVM jobs are "standard" with 2 cores or "highmem" (whatever the exact numbers work out to)?

Reply:
Ah, we should just make this the correct memory. Where can I get that from the worker?

Reply:
`job_spec['resources']['memory_bytes']`

Reply:
OK, I do not start JVMs except on standard or D